[distributed] allow rpc timeout to be set from python for different rpc backends

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29068 [distributed] allow rpc timeout to be set from python for different rpc backends**
* #28392 [distributed] add support for rpc timeouts in process group agent

RPC timeouts should be set from python for all backends, not just for process group agent

Differential Revision: [D18148449](https://our.internmc.facebook.com/intern/diff/D18148449/)

rohan-varma committed:
[distributed] allow rpc timeout to be set from python for different rpc backends

RPC timeouts should be set from python for all backends, not just for process group agent

Differential Revision: [D18148449](https://our.internmc.facebook.com/intern/diff/D18148449/)

[ghstack-poisoned]

xush6528 commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -45,6 +45,7 @@ def process_group_init_backend_handler(
     self_rank,
     worker_name_to_id,
     num_send_recv_threads,
```
Can we remove `num_send_recv_threads ` as it's not common to all RPC agents at the moment.

rohan-varma replied:
This is a function specifically for initializing PG agent though, wouldn't it be better to have the argument explicitly?

xush6528 replied:
Yeah, we can make it explicit. I think I am only hoping this argument not showing up in the arg list of `init_model_parallel`.

I think it's good now.

mrshenli replied:
We can move `num_send_recv_threads` to `backend_config` once we have updated the `init_model_parallel` API. 

+1 it's good for now

xush6528 replied:
@mrshenli 
No worry here. `num_send_recv_threads ` will be put into the `RpcAgentOptions` in my coming PR.

xush6528 commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -76,7 +77,7 @@ def process_group_init_backend_handler(
             )
         # TODO: add try-except and destroy _agent in all processes if any fails.
-        return dist.ProcessGroupAgent(self_name, group, num_send_recv_threads, kwargs["rpc_timeout"])
+        return dist.ProcessGroupAgent(self_name, group, num_send_recv_threads, rpc_timeout)
```
```cpp
return dist.ProcessGroupAgent(self_name, group, kwargs["num_send_recv_threads"], rpc_timeout)
```

rohan-varma replied:
This reads a bit odd as we are passing args from `kwargs` before regular arguments. If we want to do this, then we should change the argument order in process group agent?

rohan-varma committed:
Update on "[distributed] allow rpc timeout to be set from python for different rpc backends"


RPC timeouts should be set from python for all backends, not just for process group agent

Differential Revision: [D18148449](https://our.internmc.facebook.com/intern/diff/D18148449/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] allow rpc timeout to be set from python for different rpc backends"


RPC timeouts should be set from python for all backends, not just for process group agent

Differential Revision: [D18148449](https://our.internmc.facebook.com/intern/diff/D18148449/)

[ghstack-poisoned]

pietern commented:
This needs a rebase now that #27529 is merged.

See https://github.com/pytorch/pytorch/commit/b4df413712098ff93da694472d95896e62faf07d#diff-d0c3a068b64bd33cff798988a76be111.

xush6528 commented:
Just noticed #29336 covers this PR. I am trying to unlock myself there.

xush6528 commented:
@rohan-varma  Can we close this PR, given #29341 is ready?
